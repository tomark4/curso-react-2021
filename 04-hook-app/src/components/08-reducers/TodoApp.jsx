import { useEffect, useReducer } from 'react';
import todoReducer from './todoReducer';
import './reducers.css';
import TodoItem from './TodoItem';
import TodoAdd from './TodoAdd';

const init = () => {
    return JSON.parse(localStorage.getItem('todos')) || [];
}

const TodoApp = () => {
    const [todos, dispatch] = useReducer(todoReducer, [], init);

    useEffect(() => {
        localStorage.setItem('todos',JSON.stringify(todos));
    },[todos])

    const handleDelete = (todoId) => {
        dispatch({
            type: 'DELETE',
            payload: todoId
        });
    }

    const handleToggle = (todoId) => {
        dispatch({
            type: 'TOGGLE',
            payload: todoId
        });
    }

    const handleAddTodo = (newTodo) => {
        dispatch({
            type: 'ADD',
            payload: newTodo
        });
    }

    return (
        <div className="container my-5">
            <h3>TODO app</h3>
            <hr/>
            <p className="my-2">todos en mi lista: { todos.length }</p>
            <div className="row">
                <div className="col-md-8">
                    {todos.length === 0 &&
                        <div className="alert alert-info">No hay todos</div>
                    }
                    <ul className="list-group list-group-flush">
                        {todos.map( todo => (
                            <TodoItem todo={todo}
                            key={todo.id}
                            onDelete={handleDelete}
                            onTouch={handleToggle}
                            />
                        ))}
                    </ul>
                </div>
                <div className="col-md-4">
                    <TodoAdd onAdd={handleAddTodo} />
                </div>
            </div>
        </div>
    )
}

export default TodoApp
