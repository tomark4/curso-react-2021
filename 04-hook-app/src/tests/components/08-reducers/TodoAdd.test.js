import TodoApp from '../../../components/08-reducers/TodoAdd';
import {shallow} from 'enzyme';

describe('pruebas en todo app', () => {
  const handleAdd = jest.fn();

  const wrapper = shallow(<TodoApp onAdd={handleAdd}/>)  ;

  test('should retornar el componente', () => {
      expect(wrapper).toMatchSnapshot();
  });

  test('should no debe de llamar onAdd', () => {
    const formSubmit = wrapper.find('form').prop('onSubmit');

    formSubmit({preventDefault(){}})

    // esperaba que fuera llamado 0 veces
    expect(handleAdd).toHaveBeenCalledTimes(0)
  });

  test('should debe de llamar la funcion add todo', () => {

    const event = {
      target:{
        name: 'name',
        value: 'prueba'
      }
    };

    wrapper.find("input").simulate('change', event);
    const formSubmit = wrapper.find('form').prop('onSubmit');
    formSubmit({ preventDefault(){} });
    expect( handleAdd ).toHaveBeenCalledTimes(1);
    expect( handleAdd ).toHaveBeenCalledWith(expect.any(Object));
    expect( handleAdd ).toHaveBeenCalledWith( {id: "", name: 'prueba', done: false} );
  });

})
