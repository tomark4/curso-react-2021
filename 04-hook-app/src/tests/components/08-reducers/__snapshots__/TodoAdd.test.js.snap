// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en todo app should retornar el componente 1`] = `
<Fragment>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <label
        htmlFor=""
      >
        Name
      </label>
      <input
        autoComplete="off"
        className="form-control"
        name="name"
        onChange={[Function]}
        placeholder="Task"
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <button
        className="btn btn-outline-primary btn-block"
        type="submit"
      >
        Save
      </button>
    </div>
  </form>
</Fragment>
`;
